ocaml/forkexecd/lib/forkhelpers.ml

```diff
@@ -62,12 +62,9 @@ let waitpid (sock, pid) =
 
 let waitpid_nohang ((sock, _) as x) =
   Unix.set_nonblock sock ;
-  let r =
-    try waitpid x
-    with Unix.(Unix_error ((EAGAIN | EWOULDBLOCK), _, _)) ->
-      (0, Unix.WEXITED 0)
-  in
-  Unix.clear_nonblock sock ; r
+  try waitpid x
+  with Unix.(Unix_error ((EAGAIN | EWOULDBLOCK), _, _)) ->
+    Unix.clear_nonblock sock ; (0, Unix.WEXITED 0)
```

Comment:
This signals that the the child terminated. But don't we have to close the socket here? This is the behavior of `waitpid`. Or where and when would the socket be closed?

Reply:
I think we may have more leaks in the callers, because the callers are not consistent how they use it:
stunnel calls this, and if the pid is still alive then sends a SIGKILL and then leaks the socket.
OTOH the xapi host backup code calls waitpid_nohang in a loop with a delay inbetween, so in that case if the process is still alive we must *not* close the socket.

It might be better to replace this function with 2:
* one which always explicitly closes the socket no matter what, and doesn't get blocked if the PID is still running, e.g. 'waitpid_nohang_close'
* one which leaves the socket open so it can be called in a loop, e.g. 'waitpid_nohang_leaveopen'

Could also try to wrap it with my safe resources code, but that will only detect leaks at runtime, not compile time.